Support Interactions

JustCat80 opened the pull request:
Recreated from #1210 for hacktober

## Notes
https://discord.com/developers/docs/interactions/receiving-and-responding#interactions
https://discord.com/developers/docs/interactions/application-commands

Auto Updating Docs (built by @DonovanDMC)
https://eris.owo-whats-this.dev/interactions/

Adds support for Interactions, which adds the ability to reply to [Application Commands](https://discord.com/developers/docs/interactions/application-commands), [Message Components](https://discord.com/developers/docs/interactions/message-components), and other related things.
Also adds support for [Application Commands](https://discord.com/developers/docs/interactions/application-commands)
~~For Message Components, check out bsian's pr #1189~~

Please, let me know of any Criticisms you may have and/or changes that should be made.
### Status of being merged:
~~_As per [Abal's comment on slash commands](https://github.com/abalabahaha/eris/issues/1100#issuecomment-812851950): 
"Eris doesn't plan to implement full support for slash commands until the API gets improved+stabilized."_~~

**As per [this discord post](https://support-dev.discord.com/hc/en-us/articles/4404772028055) this is now required soon because as of april 2022 message content will become a privileged intent, and will likely be merged when discord slows down the rate of changes**
## TODO:
Fix https://github.com/abalabahaha/eris/pull/1210#issuecomment-933005417
Add WebhookClient
## Additions:
_updated as of commit 9f1d485 and may not reflect exact changes of commits after it_
### Events:
- Client Event: interactionCreate
#### Structures:
- Interaction
- PingInteraction
- CommandInteraction
- ComponentInteraction
- AutocompleteInteraction
- UnknownInteraction
#### PingInteraction Methods:
- acknowledge
- pong
#### CommandInteraction Methods:
- acknowledge
- createFollowup
- createMessage
- defer
- deleteMessage
- deleteOriginalMessage
- editMessage
- editOriginalMessage
- getOriginalMessage
#### ComponentInteraction Methods:
- acknowledge
- createFollowup
- createMessage
- defer
- deferUpdate
- deleteMessage
- deleteOriginalMessage
- editMessage
- editOriginalMessage
- editParent
- getOriginalMessage
#### AutocompleteInteraction Methods:
- acknowledge
- result
#### UnknownInteraction Methods:
UnknownInteraction has all the methods of the Interaction types listed above, just in case you know what will work
#### Client Methods:
- createInteractionResponse
- bulkEditCommandPermissions
- bulkEditCommands
- bulkEditGuildCommands
- createCommand
- createGuildCommand
- deleteCommand
- deleteGuildCommand
- editCommand
- editCommandPermissions
- editGuildCommand
- getCommand
- getCommandPermissions
- getCommands
- getGuildCommand
- getGuildCommandPermissions
- getGuildCommands
#### Guild Methods:
- bulkEditCommandPermissions
- bulkEditCommands
- createCommand
- deleteCommand
- editCommand
- editCommandPermissions
- getCommand
- getCommandPermissions
- getCommands
- getGuildCommandPermissions
#### Constants:
- new MessageFlags value
- InteractionTypes
- InteractionResponseTypes
- CommandOptionTypes
- CommandPermissionTypes
- CommandTypes
- ButtonStyles
#### Misc:
- [Endpoints](https://github.com/abalabahaha/eris/blob/1833fcf/lib/rest/Endpoints.js)
- ts docs

Commit:
Added interaction event, Added Interaction class (1)

Commit:
Fixed accidental commas

Commit:
Update Constants.js

Commit:
Temp commented out InteractionData to test so far

Commit:
Update Interaction.js

Commit:
Added endpoints

Commit:
Forgot / before urls

Commit:
Update Interaction.js

Commit:
Update Interaction.js

Commit:
Update Interaction.js

Commit:
Update Interaction.js

Commit:
Update Interaction.js

Commit:
Update Interaction.js

Commit:
Update Interaction.js

Commit:
Update Interaction.js

Commit:
Update Client.js

Commit:
Update Endpoints.js

Commit:
Update Interaction.js

Commit:
Update Interaction.js

Commit:
Update Interaction.js

Commit:
Update Interaction.js

Commit:
Interaction return changed

Commit:
Attempt to add Interaction Data

Commit:
Update Interaction.js

Commit:
Removed InteractionData.customId

Commit:
Actually did InteractionData Docs

Commit:
acknowledge is just a promise

Commit:
Defer test

Commit:
editResponse test

Commit:
Update Interaction.js

Commit:
Update Endpoints.js

Commit:
Update Endpoints.js

Commit:
Update Interaction.js

Commit:
editResponse Documentation

Commit:
Added testing for content

Commit:
Removed duplicate endpoints

Commit:
Added createFollowup

Commit:
Fixed acknowledge

Commit:
Update Interaction.js

Commit:
Update Interaction.js

Commit:
Test log

Commit:
Commented out broken .then statement

Commit:
Changed some docs, added .edit

Commit:
Update Interaction.js

Commit:
edit => editParent, editResponse => edit

Commit:
Removed editInteractionMessage

Commit:
Updated Docs

Commit:
Update Interaction.js

Commit:
Update Interaction.js

Commit:
Added interaction.delete

Commit:
Update Interaction.js

Commit:
Update Interaction.js

Commit:
Removed dead space between comments and functions

Commit:
Update Interaction.js

Commit:
Added Name to InteractionData

Commit:
Added Requested changes

• Linted
• Added createInteractionResponse client method
• Added flags to executeWebhook docs
• started using client methods in Interaction
• No longer assign null for fields
• Now checks fields for undefined explicitly
• Fixed Constants Styling (i think?)
• Removed InteractionData class
• Added Typescript documentation
• Added .vscode to .gitignore

Commit:
Removed _client that i forgot to

Commit:
fixed small docs bit

for some reason .request in interactionCreate is not defined so imma try and fix that but i see nothing wrong

Commit:
Added this._client to all method calls, everything now actually works

Commit:
Fixed Interaction docs

Commit:
added flags to executewebhook for interactions

Commit:
added createInteractionResponse to ts docs

Commit:
Fixed some docs

Commit:
Made interaction functions docs InteractionOptions rather then Record<string, unknown>

Not exactly sure how this works but maybe this im sorry if not eeeeeee

Commit:
Added missing client for member

Commit:
Added forgotten guild object

Commit:
Better docs + notes

Commit:
Updated some docs

Commit:
Added createFollowup options

Commit:
Fixed up some ts docs

Commit:
Fixed delete docs, fixed needing embed when its embeds

Commit:
Moved docs a little

Commit:
Components-only for editParent and edit

Commit:
Dont format allowedMentions because editWebhookMessage does right after

Commit:
then => than in interaction docs

JustCat80 committed:
Merge pull request #17 from abalabahaha/dev

Pull dev changes

Commit:
Added forgotten property, fixed docs a little

Commit:
Added forgotten consistent, actually follows styling now

Commit:
Added endpoints, Added consistents

Commit:
Id => ID

Commit:
Add more endpoints

Commit:
Add most client slash command functions

Still missing all Permissions related ones

Commit:
Fixed slash command docs

Commit:
Added most guild slash command functions

JustCat80 committed:
Merge pull request #19 from abalabahaha/dev

Catchup dev updates

JustCat80 committed:
Merge pull request #20 from abalabahaha/dev

Catchup dev updates

Commit:
bulkEditCommands

createCommands => bulkEditCommands

createGuildCommands => bulkEditGuildCommands

Commit:
Merge branch 'slash-commands' of https://github.com/Catboi8/eris into slash-commands

Commit:
Fixed "will will" typo

Commit:
Removed unneeded if statements, embed => embeds, fixed docs

Commit:
make interaction.edit messageId docs more clear

JustCat80 committed:
Merge pull request #26 from abalabahaha/dev

Pull dev updates

Commit:
Merge branch 'dev' of https://github.com/abalabahaha/eris into abalabahaha-dev

Commit:
Merge branch 'abalabahaha-dev' into interactions

JustCat80 committed:
Merge pull request #30 from abalabahaha/dev

Dev updates

Commit:
docs improvements

Constants.InteractionResponseType => Constants.InteractionResponseTypes
Moved ts docs to the same spot
Added more missing ts docs
Fixed some js docs

Added "Errors with ephemeral messages." to editParent

Fix some docs

Fix and improve ts docs
Fix createInteractionResponse docs

Commit:
Ignore .vscode

Commit:
Fixed some defer docs, acknowledge now has a reason to exist

Acknowledge now works differently depending on the interaction type,
Works as a Pong when used on a type 1 interaction. (Ping)
Works as defer() when used on a type 2 interaction. (Slash Command)
Works as deferUpdate() when used on a type 3 interaction. (Message Comonent)

Commit:
Add Donovan's suggested changes, also values is now an array of strings

Co-Authored-By: Donovan Daniels <hewwo@yiff.rocks>

Commit:
embeds is now an actual array of EmbedOptions

Co-Authored-By: Reinhardt <notmarx.tech@gmail.com>

Commit:
Remove Guild#deleteGuildCommand, fix some docs

Commit:
Linted ts

Commit:
Remove unrelated endpoints

Commit:
Added interaction.getOriginal() https://discord.com/developers/docs/interactions/slash-commands#get-original-interaction-response

Commit:
change function names to make more sense, ts docs

getOriginal => getOriginalMessage
edit => editMessage
delete => deleteMessage

Commit:
"Warning: Will error with ephemeral messages." is now put below the notes

Commit:
slight jsdoc improvements

flags was a Boolean in the jsdocs, when it is a Number
slightly change wording

Commit:
(see description) Added deleteOriginalMessage and editOriginalMessage

I will remove these if asked to, I added them because of getOriginalMessage being a thing, let me know if you think these shouldn't stay

Commit:
Id => ID

Commit:
Fix some docs

Commit:
Added Permission Functions to client and guild

- Client#getGuildCommandPermissions
- Client#getCommandPermissions
- Client#editCommandPermissions
- Client#bulkEditCommandPermissions

- Guilld#getGuildCommandPermissions
- Guilld#getCommandPermissions
- Guilld#editCommandPermissions
- Guilld#bulkEditCommandPermissions

remove trailing spaces

Commit:
Added suggested docs changes

Co-Authored-By: Reinhardt <notmarx.tech@gmail.com>

Commit:
Added CommandTypes to Constants and command.type to jsdocs

Note: Currently in private beta

Commit:
Added ts docs :)

Will need to be double checked, I'm not the greatest at ts

Commit:
Oops, linted ts

Commit:
Added NUMBER (type 10) to Constants.CommandOptionTypes

JustCat80 committed:
Update index.d.ts

Co-authored-by: DetachHead <57028336+DetachHead@users.noreply.github.com>

Commit:
Added new Interaction types that extend Interaction, added ts docs, changed Constants, added some missing js and tsdocs

PingInteraction (type 1)
CommandInteraction (type 2)
ComponentInteraction (type 3)
UnknownInteraction (any types in the future eris doesn't support yet)

Commit:
fix ts docs

data isn't optional for CommandInteraction and ComponentInteraction
componentType isn't optional for ComponentInteraction
data is unknown for UnknownInteraction
type is numberfor UnknownInteraction

Commit:
for...in => for...of

Co-Authored-By: DetachHead <57028336+DetachHead@users.noreply.github.com>

Commit:
Merge branch 'slash-commands' into combine

Commit:
Merge branch 'combine' into interactions

Commit:
Forgot Linting

Commit:
Added requested changes

now use keyof
added classes to exports
changed interaction type to number

Co-Authored-By: Donovan Daniels <hewwo@yiff.rocks>

Commit:
Added more requested changes

now use Record
componentType is now component_type
choices is an array

Commit:
Assign wait to all createFollowup methods

JustCat80 committed:
Change value back to string

Co-authored-by: Donovan Daniels <hewwo@yiff.rocks>

Commit:
Fix Guild#bulkEditCommands

JustCat80 committed:
Merge pull request #46 from Catboi8/interactions

Interactions updated

Commit:
Added Interaction.acknowledged, update Interaction.acknowledged when using a function that a acknowledges the interaction

Commit:
acknowledged docs

JustCat80 committed:
Push Acknowledged consept to main interactions branch

Push Acknowledged consept to main interactions branch

Commit:
Lint ts

JustCat80 committed:
Update lib/Client.js

Co-authored-by: Donovan Daniels <hewwo@yiff.rocks>

Commit:
Added file to createMessage docs because after the initial response you can send a file

Commit:
Docs changes, Constants changes

Change mentions of Slash Command to Application Command unless directly referencing the Slash Command Application Command type

Commit:
Changed Constants COMMAND to CHAT_INPUT, changed name check

TTtie committed:
use raw messages when interaction messages are ephemeral

TTtie committed:
use a bit mask instead of direct comparison

TTtie committed:
use an explicit 64 for PossiblyEphemeralMessage.flags

JustCat80 committed:
Merge pull request #52 from TTtie/ephemeral-message-passthrough

Commit:
Removed InteractionFlags in favor of adding EPHEMERAL to MessageFlags, fixed some constants styling

Commit:
Application Commands Example

JustCat80 committed:
Fix message context menu comment

Co-authored-by: Loliticos <47262825+Loliticos@users.noreply.github.com>

Commit:
editOriginalMessage doesnt error with ephemeral messages

i thought it did but maybe im dumb

Commit:
Warn if no intents instead of allowing it to error, as Interactions dont need an intent

Emit a warning and set to intents 0 if none are specified

Commit:
Change applicationCommands example

Commit:
Oops added an underscore to BOT TOKEN

Commit:
Update Constants Names to be consistent

Commit:
Added docs for forgotten context menu data

Commit:
remove PossiblyEphemeralMessage from ts docs

Commit:
Fix ComponentInteraction#message docs, make most message methods check if the message is ephemeral before attempting the method

Commit:
CONTEXT_MENU_COMMAND message type

Commit:
APPLICATION_COMMAND => CHAT_INPUT_COMMAND

Add forgotten `case MessageTypes.CONTEXT_MENU_COMMAND` to Message.js

Commit:
data => info in interaction classes, data.resolved fixed

Commit:
Oops, fix redefining things for no reason

Commit:
remove data.resolved for now

Commit:
remove unused things for now

Commit:
Merge branch 'dev' of https://github.com/abalabahaha/eris into interactions

Commit:
Add components to jsdocs

Commit:
interaction.data.resolved is now actually set instead of just passing the raw data

Commit:
Fix PartialChannel and PartialRole having id set as number

Commit:
fixed docs to reflect that data.resolved are Collections

Commit:
add APPLICATION_COMMAND events, will maybe add an ApplicationCommand class later

Commit:
Revert commit cbe15029c223b20f73ae1a07905d9769f0126819 as these events are not sent to bots

Revert "add APPLICATION_COMMAND events, will maybe add an ApplicationCommand class later"
This reverts commit cbe15029c223b20f73ae1a07905d9769f0126819.

Commit:
fix message TypeError (hotfix by GweeKwee)

fix TypeError: Cannot read property 'members' of undefined at new Message

Co-Authored-By: Gwee&Kwee <gwee@nekotina.com>

Commit:
interaction.channelID is now interaction.channel, similar to message

Commit:
fix up message.interaction, make message use this._client so its consistent

Commit:
Merge branch 'dev' into interactions

Commit:
use info.guild_id when defining members

Commit:
Update UnknownInteraction.message

Commit:
Cache CommandInteraction.resolved users and members, cache interaction.member and interaction.user

Commit:
Change options.value to string | number | boolean for now

Co-Authored-By: coolcalcacol <31094492+coolcalcacol@users.noreply.github.com>

DonovanDMC committed:
first attempt

DonovanDMC committed:
add stuff to unions

JustCat80 committed:
Merge pull request #61 from DonovanDMC/option-types

make InteractionDataOptions strictly defined

Commit:
Convert to interfaces

JustCat80 committed:
Add components to webhook content in ts

Co-authored-by: Donovan Daniels <hewwo@yiff.rocks>

Commit:
Add Interaction classes to esm.mjs

Commit:
Add ComponentTypes to constants because bsian didn't

:)

Co-Authored-By: Donovan Daniels <hewwo@yiff.rocks>

DonovanDMC committed:
Even More Detailed Interaction Types (#63)

* ComponentTypes constants

* ComponentInteraction*Data types

* use constants for type

* space

* More specific typings for `ApplicationCommandOptions`

* forgot to change stuff when I copied

* fixes and changes

* even more messy types

* correct choices types

* lint

* specific depth limits
docs: https://discord.com/developers/docs/interactions/application-commands#subcommands-and-subcommand-groups

* `ApplicationCommand#description` ternary

* these have been resolved

* lint

* even more

Commit:
ts hotfix

Commit:
remove UnknownInteraction

Commit:
Undo last commit

Commit:
export Interaction

Commit:
data.interaction.guild_id => data.guild_id in Message

Commit:
Can send files on initial interaction response, add .then update to UnknownInteraction methods

Question life because why is executeWebhook file different from createMessage file wtf

DonovanDMC committed:
More Fixes (#64)

* I suck

* same thing here

Commit:
Fix Error: Only iterables may be used in a SetType (Catboi8#68)

Fix allowed_mentions in interaction responses

Co-Authored-By: Gwee&Kwee <gwee@nekotina.com>

Commit:
Added ephemeral to Attachment ts

Commit:
Fix editParent typings

JustCat80 committed:
Fix bulkEditCommandPermissions typings

Co-authored-by: Lava <service.github@lava.moe>

Commit:
Add ts fixes, add channel_types to ApplicationCommandOptionsChannel

Co-Authored-By: Lava <me@lava.moe>

Commit:
Merge branch 'interactions' of https://github.com/Catboi8/eris into interactions

Commit:
Oops

Commit:
fix perms again xD

Co-Authored-By: Lava <me@lava.moe>

Commit:
change all acknowledges, make PingInteraction consistent

Commit:
Autocomplete Interaction Support 🥳

Add support for interaction type APPLICATION_COMMAND_AUTOCOMPLETE (4)
add unknown to ComponentInteraction data
fix InteractionContent
add acknowledge to UnknownInteraction (is just createInteractionResponse)

Commit:
Merge branch 'dev' of https://github.com/abalabahaha/eris into interactions

Commit:
Fix js docs maybe

DonovanDMC committed:
a few fixes/improvements (#69)

* start

* remove this unknown garbage

JustCat80 committed:
ComponentTypes Lint

Co-authored-by: Avocado <43632131+Awoocado@users.noreply.github.com>

JustCat80 committed:
Update index.d.ts

Co-authored-by: Reinhardt <notmarx.tech@gmail.com>

JustCat80 committed:
Update index.d.ts

Co-authored-by: Reinhardt <notmarx.tech@gmail.com>

JustCat80 committed:
Add ButtonStyles to Constants

*glares at bsian*

JustCat80 committed:
Components example

Because explaining it is meh, direct em here instead :3

JustCat80 committed:
Oops forgot intents

JustCat80 committed:
Merge branch 'dev' of https://github.com/abalabahaha/eris into interactions

JustCat80 committed:
Change message.interaction

Not sure if this is the best way to handle it but hopefully it is

JustCat80 committed:
Fix some js docs

DonovanDMC commented on index.d.ts:
```diff
@@ -1709,18 +1903,23 @@ declare namespace Eris {
     createGuildFromTemplate(code: string, name: string, icon?: string): Promise<Guild>;
     createGuildSticker(guildID: string, options: CreateStickerOptions, reason?: string): Promise<Sticker>;
     createGuildTemplate(guildID: string, name: string, description?: string | null): Promise<GuildTemplate>;
+    createInteractionResponse(interactionID: string, interactionToken: string, options: InteractionOptions, file: FileContent | FileContent[]): Promise<void>;
```
File is marked as required
```suggestion
    createInteractionResponse(interactionID: string, interactionToken: string, options: InteractionOptions, file?: FileContent | FileContent[]): Promise<void>;
```

JustCat80 committed:
Update index.d.ts

Co-authored-by: Donovan Daniels <hewwo@yiff.rocks>

JustCat80 committed:
Separate file into its own argument on interaction things

Warning: breaking changes :p

also update UnknownInteraction acknowledged warning

bsian03 committed:
Merge branch 'dev' into interactions

Awoocado commented on lib/Constants.js:
```diff
@@ -208,7 +208,8 @@ module.exports.MessageFlags = {
     SOURCE_MESSAGE_DELETED: 1 << 3,
-    URGENT:                 1 << 4
+    URGENT:                 1 << 4,
+    EPHEMERAL:              1 << 6
```
```suggestion
    EPHEMERAL:              1 << 6,
    LOADING:                1 << 7
```

Awoocado commented on index.d.ts:
```diff
@@ -1281,6 +1425,7 @@ declare namespace Eris {
       SUPPRESS_EMBEDS: 4;
       SOURCE_MESSAGE_DELETED: 8;
       URGENT: 16;
+      EPHEMERAL: 64;
```
```suggestion
      EPHEMERAL: 64;
      LOADING: 128;
```

bsian03 committed:
Add Loading flag

Co-authored-by: Avocado <43632131+Awoocado@users.noreply.github.com>

abalabahaha committed:
lint

bsian03 reviewed (approved):
yolo
